src/udev/udev-node.c

```diff
@@ -130,110 +172,394 @@ static int link_find_prioritized(sd_device *dev, bool add, const char *stackdir,
         char name_enc[PATH_MAX];
         const char *id_filename;
-        int r;
+        static int semid = SEMID_UNSET;
+        int dfd, priority;
+        unsigned short semidx;
+        int r, lock_failed;
 
         assert(dev);
         assert(slink);
 
+        if (semid == SEMID_UNSET) {
+                semid = init_link_semaphores(LINKS_DIRNAME);
+                if (semid < 0) {
+                        log_error_errno(semid, "Locking under "LINKS_DIRNAME" is disabled: %m");
+                        semid = SEMID_BAD;
+                }
+        }
+        (void) mkdir_p(LINKS_DIRNAME, 0755);
+        links_fd = open(LINKS_DIRNAME, O_RDONLY|O_DIRECTORY);
+        if (links_fd == -1)
+                return log_error_errno(errno, "Failed to open %s: %m", dirname);
+
         r = device_get_id_filename(dev, &id_filename);
         if (r < 0)
                 return log_device_debug_errno(dev, r, "Failed to get id_filename: %m");
 
         util_path_encode(slink + STRLEN("/dev"), name_enc, sizeof(name_enc));
-        dirname = path_join("/run/udev/links/", name_enc);
+        dirname = path_join(LINKS_DIRNAME, name_enc);
         if (!dirname)
                 return log_oom();
-        filename = path_join(dirname, id_filename);
-        if (!filename)
-                return log_oom();
 
-        if (!add && unlink(filename) == 0)
-                (void) rmdir(dirname);
+        r = device_get_devlink_priority(dev, &priority);
+        if (r < 0)
+                return log_device_debug_errno(dev, r, "Failed to get devlink prio: %m");
+
+        if (add) {
+                mkdirat(links_fd, name_enc, 0755);
+                dfd = openat(links_fd, name_enc, O_RDONLY|O_DIRECTORY);
+                if (dfd == -1)
+                        return log_device_error_errno(dev, errno,
+                                                      "Failed to open %s: %m", dirname);
+                create_target_entry(dfd, priority, id_filename, slink);
+        } else {
+                dfd = openat(links_fd, name_enc, O_RDONLY|O_DIRECTORY);
+                if (dfd == -1 && errno != ENOENT)
+                        return log_device_error_errno(dev, errno,
+                                                      "Failed to open %s: %m", dirname);
+                delete_target_entry(dfd, priority, id_filename, slink);
+        }
+
+        dir = fdopendir(dfd);
+        if (!dir) {
+                close(dfd);
+                return log_device_error_errno(dev, errno,
+                                              "Failed to fdopendir %d: %m", dfd);
+        }
 
-        r = link_find_prioritized(dev, add, dirname, &target);
-        if (r < 0) {
-                log_device_debug(dev, "No reference left, removing '%s'", slink);
+        semidx = get_sema_index(slink);
+        lock_failed = lock_slink(semid, semidx);
+        if (lock_failed)
+                log_error_errno(errno, "Failed to lock %s: %m", slink);
```

Comment:
`lock_slink` returns negative errno...
- `log_error_errno(lock_failed, "...");`
- Do not implicit cast to bool, so, `if (lock_failed < 0)`
- Not a matter, but log message seems to be slightly redundant, as `lock_slink()` also logs.
- Not a matter, the variable name, I prefer `ret_lock` or something instead of `lock_failed`.

Reply:
I like lock_failed because it clarifies matters further down, when I unlock. But I'll improve the code such that lock_failed is treated properly like a bool.